### PR TITLE
[Translation] Added intl message formatter.

### PR DESCRIPTION
**Pull request by aitboudad**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |

Related to #18314

### CODE REVIEW

File: src/Symfony/Component/Translation/Tests/Formatter/IntlMessageFormatterTest.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        return array(
+            array(
+                'There is one apple',
+                'There is one apple',
+                array(),
+            ),
+            array(
+                '4,560 monkeys on 123 trees make 37.073 monkeys per tree',
+                '{0,number,integer} monkeys on {1,number,integer} trees make {2,number} monkeys per tree',
```
Comment (unkind) [inline src/Symfony/Component/Translation/Tests/Formatter/IntlMessageFormatterTest.php]: What's the point in `MessageFormatter` abstraction if at the end you are coupled with specific implementation?

**Comment by Koc** [comment]
ref #10671, #10152, #6009

**Comment by fabpot** [comment]
This only makes sense if we support the plurals as well.

**Comment by aitboudad** [comment]
@fabpot I'm not sure what you mean, the [icu messageformat](http://userguide.icu-project.org/formatparse/messages) support the plurals as well.

**Comment by MatTheCat** [comment]
This seems abandoned, does this mean we'll have to add an `IntlMessageFormatter` ourselves in order to use `\MessageFormatter`?

**Comment by nicolas-grekas** [comment]
Moving to 4.1. Rebase on master needed, where PHP 7.1 features can be used btw.

**Comment by stof** [comment]
The issue with the PR in its current state is that it creates a community split: as soon as you configure Symfony to use the intl formatter in your project, it becomes incompatible with all bundles written using the default formatter (and this includes translations shipped in the core for validation errors). This makes it impossible to use it in a gradual way.

**Comment by sstok** [comment]
@stof indeed 😞 I guess the only way to resolve this is adding an extra argument to the trans method (CustomTransformerTranslatorInterface with an overwrite of trans() to add an additional argument) 🤔 not the nicest way but at least it would make is possible to use the new Transformer.

And we can't a flag to a translation file because YAML and PHP for example use a hash at the root level, and making `@format_transformer` a special key feels a bit strange 😛 and would still require a proper solution for XLIFF.

**Comment by stof** [comment]
@sstok requiring code *using* the translations to provide the formatter to use to format this message looks really weird, because it puts this info in the wrong place. The template doing a translation calls has no idea which formatter implementation the original file was written for.

> This only makes sense if we support the plurals as well.

The ICU message format supports pluralization based on any parameter (and on multiple ones), as this is one of the available features inside the string. We don't need a specific `transChoice` feature in this case (which is an API limited to a single integer being the pluralization source)

**Comment by aitboudad** [comment]
I think grouping messages by formatter would solve the issue, something like `domain` part.
I haven't yet to time to investigate on it, so if anyone interested on it would be welcomed :smile:

**Comment by fabpot** [comment]
Closing in favor of #27399

**Pull request closed without merging**